[Merged by Bors] - TY-2472 test user reacted event

Robert-Steiner opened the pull request:
Ticket:

- [TY-2472]

based on #210 

[TY-2472]: https://xainag.atlassian.net/browse/TY-2472?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

janpetschexain commented on discovery_engine/test/integration/user_reacted_test.dart:
```diff
@@ -0,0 +1,121 @@
+    });
+
+    test('change the user reaction of a document', () async {
+      server = await LocalNewsApiServer.start();
+      var engine = await DiscoveryEngine.init(
+        configuration: createConfig(data, server.port),
+      );
```
this pattern repeats over several tests, should we encapsulate this in a utility function as well? (but only if this doesn't make too much work, as we don't have a native tuple type to return in dart)

Robert-Steiner replied:
Good idea :+1 

I also moved the server + engine (when possible) into the setup function and call `engine.dispose()` at the end.
Please have another look

janpetschexain reviewed (approved):
lgtm

janpetschexain commented on discovery_engine/test/integration/utils/helpers.dart:
```diff
@@ -15,7 +15,12 @@
 import 'package:xayn_discovery_engine/discovery_engine.dart'
-    show Manifest, createManifestReader, Configuration, FeedMarket;
+    show
+        Configuration,
+        DiscoveryEngine,
+        FeedMarket,
+        Manifest,
+        createManifestReader;
```
it's weird that dart complains about unsorted `import`s, but not about unsorted `show`s.
```suggestion
        Configuration,
        createManifestReader,
        DiscoveryEngine,
        FeedMarket,
        Manifest;
```

janpetschexain commented on discovery_engine/test/integration/init_engine_test.dart:
```diff
@@ -40,23 +41,17 @@ void main() {
-
+      final engine = await initEngine(data, server.port);
       expect(engine, isA<DiscoveryEngine>());
+      await engine.dispose();
```
this line could be moved to the `tearDown()` like in the other tests. similarly below.

janpetschexain commented on discovery_engine/test/integration/init_engine_test.dart:
```diff
@@ -40,23 +41,17 @@ void main() {
-        configuration: createConfig(data, server.port),
-      );
-
+      final engine = await initEngine(data, server.port);
```
this line could be moved to the `setUp()` like in the other tests. similarly below.

Robert-Steiner replied:
unfortunately, that will break the second test case. it will pass but we don't check if the initialisation is   successful in case of an api error . but we can move the dispose into `tearDown()`

janpetschexain reviewed (commented):
lgtm. a test has timed out though, is it enough to increase the timeout or is this due to some external factors?

Robert-Steiner commented:
> lgtm. a test has timed out though, is it enough to increase the timeout or is this due to some external factors?

hm, that timeout is not so good. i also got that error locally after i rebased and added the `engine.dispose()` calls. I thought that happened because I was running the test while i was alter them. I will take a closer look

Robert-Steiner commented:
it would be possible that we spawn too many engines at the same time and thus everything is slowed down.

Robert-Steiner commented:
the changeUserReaction request takes on my machine around 9.5 seconds and out timeout is 10 seconds. maybe we should think about comiling the rust binary in release mode to speed things up

Robert-Steiner commented:
> the changeUserReaction request takes on my machine around 9.5 seconds and our timeout is 10 seconds. maybe we should think about compiling the rust binary in release mode to speed things up

we can also just change the opt-level of `tract-core` and `tract-data`
```
[profile.dev.package.tract-core]
opt-level = 3
[profile.dev.package.tract-data]
opt-level = 3
```
`tract-core` reduces the `time log_user_reaction` time from 15sec to 6sec
and `tract-data` reduces the time of the engine init from 9.5sec to 3sec

Robert-Steiner commented:
> Is the increase in compilation time noticeable?

yes, it takes 2 min more

what i saw is that they do 
```
[profile.dev.package."*"]
opt-level = 3
```

but this setting should not affect our build options or did i misunderstood you?

Robert-Steiner commented:
The new smbert model makes the initialisation of the engine much faster. Even without optimisations and running the tests in parallel it takes 4 sec instead of 15 sec. however `changeUserReaction` got slower from 9 sec to 14 sec (without opt, running in parallel). 

what we can do is: 
- enable `opt = 3` for `tract-core`
	- 7 sec running tests in parallel, 5 secs sequential
	- but that will take 2 min longer to compile (but it is not recompiled every time/ will be slower on the ci if we don't have cache hit) 

wdyt?

unfortunately, running the dart tests in sequentially will be to slow

Robert-Steiner commented:
like discussed in the daily we compile the tract-core crate with `opt=3` to make the dart test run faster

bors merge

Robert-Steiner commented:
bors merge
